apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js

```diff
@@ -141,6 +142,25 @@ GoogleBlockly.Extensions.register('behaviors_block_frame', function () {
   }
 });
 
+// This extension is used to update the block's behaviorId when a user-created behavior is renamed.
+// TODO: Add logic to update the dropdown options on behaviorPicker blocks too.
+GoogleBlockly.Extensions.register('behaviors_name_validator', function () {
+  const nameField = this.getField('NAME');
+  nameField.setValidator(function (newValue) {
+    // The default validator provided by mainline Blockly. Strips whitespace.
+    const rename = Blockly.Procedures.rename.bind(this);
+    const legalName = rename(newValue);
```

Comment:
if it's not a legal name I assume the block won't actually be renamed, right?

Reply:
So it seems like `Procedures.rename` basically ensures that we always end up with a legal name. It works by trimming leading/trailing white space, then ensuring it doesn't conflict with another procedure name, and finally adding a number on the end if does. If the new name is falsy (empty string) it gets replaced with a translation of the string "unnamed", provided that Blockly has one.

https://github.com/google/blockly/blob/925a7b9723ac46217c64a1842668fd0a66549449/core/procedures.ts#L193-L225

https://github.com/google/blockly/blob/925a7b9723ac46217c64a1842668fd0a66549449/core/procedures.ts#L113-L138